src/partial-evaluators/ArrayExpression.js

```diff
@@ -0,0 +1,149 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import type { BabelNodeArrayExpression } from "babel-types";
+import type { LexicalEnvironment } from "../environment.js";
+import type { Realm } from "../realm.js";
+
+import { AbruptCompletion, PossiblyNormalCompletion } from "../completions.js";
+import { TypesDomain, ValuesDomain } from "../domains/index.js";
+import { ArrayCreate, CreateDataProperty, GetIterator, GetMethod, IteratorStep, IteratorValue, Set } from "../methods/index.js";
+import { AbstractValue, NumberValue, ObjectValue, StringValue, Value } from "../values/index.js";
+
+import invariant from "../invariant.js";
+import * as t from "babel-types";
+
+// ECMA262 2.2.5.3
+export default function (
+  ast: BabelNodeArrayExpression, strictCode: boolean, env: LexicalEnvironment, realm: Realm
+): [AbruptCompletion | Value, BabelNodeArrayExpression] {
+  // 1. Let array be ArrayCreate(0).
+  let array = ArrayCreate(realm, 0);
+
+  // 2. Let len be the result of performing ArrayAccumulation for ElementList with arguments array and 0.
+  let elements = ast.elements || [];
+  let partial_elements = [];
+  let len = elements.length;
+  let nextIndex = 0;
+  for (let i = 0; i < len; i++) {
+    let elem = elements[i];
+    if (!elem) {
+      nextIndex++;
+      continue;
+    }
+
+    let elemValue, elemAst;
+    if (elem.type === "SpreadElement")
+      [elemValue, elemAst] = env.partiallyEvaluateCompletionDeref(elem.argument, strictCode);
+    else
+      [elemValue, elemAst] = env.partiallyEvaluateCompletionDeref(elem, strictCode);
+    if (elemValue instanceof AbruptCompletion) {
+      return [elemValue, ast]; //todo: log an error message
+    } else if (elemValue instanceof PossiblyNormalCompletion) {
+      // TODO: there was a conditional abrupt completion while evaluating elem, so join states somehow
+      return [AbstractValue.createIntrospectionErrorThrowCompletion(elemValue.value), ast];
+    }
+    invariant(elemValue instanceof Value);
+    partial_elements[nextIndex] = (elemAst: any);
+
+    // ECMA262 12.2.5.2
+    if (elem.type === "SpreadElement") {
+      let spreadObj = elemValue;
+      partial_elements[nextIndex] = t.spreadElement((elemAst: any));
+
+      // update the abstract state with the contents of spreadObj, if known
+      if (spreadObj instanceof ObjectValue && !spreadObj.isPartial()) {
+        // 3. Let iterator be ? GetIterator(spreadObj).
+        let iterator = GetIterator(realm, spreadObj);
+
+        // 4. Repeat
+        while (true) {
+          // a. Let next be ? IteratorStep(iterator).
+          let next = IteratorStep(realm, iterator);
```

Comment:
It works very similar to the non partial evaluator: The exception terminates this code, then gets caught and turned into a throw completion. The completion bubbles up to the containing AST, which re-throws it and so on. When a relevant catch clause is reached, the exception gets swallowed and execution proceeds. Note that this means that the state that reaches the catch clause is the state that applied when the exception was raised and that the containing catch will do the necessary joins etc.

The bulk of this still has to be written so I can't test this right now, but that is my thinking and it seems reasonable to me right now.